From d19e5d749eefccbc6a53461ebea1d0747cc3948d Mon Sep 17 00:00:00 2001
From: Evgeny Vereshchagin <evvers@ya.ru>
Date: Sun, 24 Mar 2024 17:23:35 +0000
Subject: [PATCH] core: reject more service names violating RFC6763

Accodring to https://datatracker.ietf.org/doc/html/rfc6763#section-9
```
A DNS query for
   PTR records with the name "_services._dns-sd._udp.<Domain>" yields a
   set of PTR records, where the rdata of each PTR record is the two-
   label <Service> name, plus the same domain, e.g.,
   "_http._tcp.<Domain>".
```
and according to https://datatracker.ietf.org/doc/html/rfc6763#section-7
```
The <Service> portion of a Service Instance Name consists of a pair
   of DNS labels, following the convention already established for SRV
   records [RFC2782].
```

With this patch applied avahi rejects service names violating those
rules. It prevents avahi-browse from showing one-label service names and
also prevents avahi-discover-standalone from crashing on receiving
announcements with non-compliant PTR resource records.

In the case of
avahi-discover-standalone it was technically a DoS allowing packets
with unsolicited announcements containing non-compliant service names to
crash it. In practice it was triggered by a VM running Home Assistant OS
and sending announcements with PTR resource records pointing
`_services._dns-sd._udp.local` to `_services._dns-sd._udp.local`.

Closes https://github.com/avahi/avahi/issues/581
---
 avahi-common/domain.c            | 2 +-
 avahi-core/browse-service-type.c | 5 +++++
 avahi-core/browse-service.c      | 5 +++++
 3 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/avahi-common/domain.c b/avahi-common/domain.c
index 9e93018..3c01d43 100644
--- a/avahi-common/domain.c
+++ b/avahi-common/domain.c
@@ -578,7 +578,7 @@ int avahi_service_name_split(const char *p, char *name, size_t name_size, char *
     if ((oname && !avahi_is_valid_service_name(oname)))
         return AVAHI_ERR_INVALID_SERVICE_NAME;
 
-    if (!avahi_is_valid_service_type_generic(otype))
+    if (!avahi_is_valid_service_type_strict(otype) && !avahi_is_valid_service_subtype(otype))
         return AVAHI_ERR_INVALID_SERVICE_TYPE;
 
     if (!avahi_is_valid_domain_name(odomain))
diff --git a/avahi-core/browse-service-type.c b/avahi-core/browse-service-type.c
index f0d6938..e57b8da 100644
--- a/avahi-core/browse-service-type.c
+++ b/avahi-core/browse-service-type.c
@@ -69,6 +69,11 @@ static void record_browser_callback(
             return;
         }
 
+        if (!avahi_is_valid_service_type_strict(type)) {
+            avahi_log_debug("Invalid service '%s'", record->data.ptr.name);
+            return;
+        }
+
         b->callback(b, interface, protocol, event, type, domain, flags, b->userdata);
     } else
         b->callback(b, interface, protocol, event, NULL, b->domain_name, flags, b->userdata);
diff --git a/avahi-core/browse-service.c b/avahi-core/browse-service.c
index e924bae..830659b 100644
--- a/avahi-core/browse-service.c
+++ b/avahi-core/browse-service.c
@@ -73,6 +73,11 @@ static void record_browser_callback(
             return;
         }
 
+        if (!avahi_is_valid_service_type_strict(type)) {
+            avahi_log_debug("Invalid service '%s'", record->data.ptr.name);
+            return;
+        }
+
         b->callback(b, interface, protocol, event, service, type, domain, flags, b->userdata);
 
     } else
